isa/kernel.org/instruction-set.rst

```diff
@@ -318,49 +446,161 @@ the atomics features, while keeping a lower ``-mcpu`` version, you can use
 
-Instructions with the ``BPF_IMM`` mode modifier use the wide instruction
+Instructions with the ``BPF_IMM`` 'mode' modifier use the wide instruction
 encoding for an extra imm64 value.
 
 There is currently only one such instruction.
 
 ``BPF_LD | BPF_DW | BPF_IMM`` means::
 
-  dst_reg = imm64
+  dst = imm64
 
 
 Legacy BPF Packet access instructions
 -------------------------------------
 
 eBPF has special instructions for access to packet data that have been
 carried over from classic BPF to retain the performance of legacy socket
-filters running in the eBPF interpreter.
+filters running in an eBPF interpreter.
 
 The instructions come in two forms: ``BPF_ABS | <size> | BPF_LD`` and
 ``BPF_IND | <size> | BPF_LD``.
 
 These instructions are used to access packet data and can only be used when
-the program context is a pointer to networking packet.  ``BPF_ABS``
+the program context contains a pointer to a networking packet.  ``BPF_ABS``
 accesses packet data at an absolute offset specified by the immediate data
 and ``BPF_IND`` access packet data at an offset that includes the value of
 a register in addition to the immediate data.
 
 These instructions have seven implicit operands:
 
- * Register R6 is an implicit input that must contain pointer to a
-   struct sk_buff.
+ * Register R6 is an implicit input that must contain a pointer to a
+   context structure with a packet data pointer.
  * Register R0 is an implicit output which contains the data fetched from
    the packet.
- * Registers R1-R5 are scratch registers that are clobbered after a call to
-   ``BPF_ABS | BPF_LD`` or ``BPF_IND | BPF_LD`` instructions.
+ * Registers R1-R5 are scratch registers that are clobbered by the
+   instruction.
 
-These instructions have an implicit program exit condition as well. When an
-eBPF program is trying to access the data beyond the packet boundary, the
-program execution will be aborted.
+*Linux implementation note*: In Linux, R6 references a struct sk_buff.
+
+These instructions have an implicit program exit condition as well. If an
+eBPF program attempts access data beyond the packet boundary, the
+program execution must be gracefully aborted.
 
 ``BPF_ABS | BPF_W | BPF_LD`` means::
 
-  R0 = ntohl(*(u32 *) (((struct sk_buff *) R6)->data + imm32))
+  R0 = ntohl(*(uint32_t *) (R6->data + imm))
+
+where `ntohl()` converts a 32-bit value from network byte order to host byte order.
 
 ``BPF_IND | BPF_W | BPF_LD`` means::
 
-  R0 = ntohl(*(u32 *) (((struct sk_buff *) R6)->data + src_reg + imm32))
+  R0 = ntohl(*(uint32_t *) (R6->data + src + imm))
+
+Appendix
+========
+
+For reference, the following table lists opcodes in order by value.
+
+======  =================================================  =============
+opcode  description                                        reference 
+======  =================================================  =============
+0x04    dst = (uint32_t)(dst + imm)                        `Arithmetic instructions`_
```

Comment:
In my opinion, this table is the biggest thing that has been missing, and will be most essential for a more "formal" specification.

I had put together something similar that I sent out to the bpf mailing list a few months ago:

https://lore.kernel.org/bpf/8DA9E260-AE56-4B21-90BF-CF0049CFD04D@intel.com/

I think it would be helpful to show how the opcodes are constructed as an additional column, for example 0x04 = BPF_ALU | BPF_K | BPF_ADD.